README.rst
===================================================
Community Detection iQuery Gene To Term Mapper
===================================================

This repository creates a CDAPS compatible functional enrichment Docker image using `iQuery service <http://iquery.ndexbio.org>`_

Dependencies
------------

* `Docker <https://www.docker.com/>`_
* `make <https://www.gnu.org/software/make/>`_ (to build)
* Python (to build)

Direct invocation
------------------

Version `0.4.0` can be directly pulled from `Dockerhub <https://hub.docker.com/>`_ with this command:

.. code-block::

   docker pull coleslawndex/cdiquerygenestoterm:0.4.0

Building
--------

.. code-block::

   git clone https://github.com/idekerlab/cdiquerygenestoterm
   cd cdiquerygenestoterm
   make dockerbuild

Run **make** command with no arguments to see other build/deploy options including creation of Docker image

.. code-block::

   make

Output:

.. code-block::

   clean                remove all build, test, coverage and Python artifacts
   clean-build          remove build artifacts
   clean-pyc            remove Python file artifacts
   clean-test           remove test and coverage artifacts
   lint                 check style with flake8
   test                 run tests quickly with the default Python
   test-all             run tests on every Python version with tox
   coverage             check code coverage quickly with the default Python
   docs                 generate Sphinx HTML documentation, including API docs
   servedocs            compile the docs watching for changes
   testrelease          package and upload a TEST release
   release              package and upload a release
   dist                 builds source and wheel package
   install              install the package to the active Python's site-packages
   dockerbuild          build docker image and store in local repository
   dockerpush           push image to dockerhub


Usage
-----

.. code-block::

   docker run -v coleslawndex/cdiquerygenestoterm:0.4.0 -h


Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.4.0 (2020-03-06)
------------------

* Code now returns hit with highest similarity score instead of by pvalue

0.1.0 (2019-09-19)
------------------

* First release on PyPI.


tests/test_cdiquerygenestoterm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_cdgprofilergenestoterm
----------------------------------

Tests for `cdgprofilergenestoterm` module.
"""

import os
import sys
import unittest
import tempfile
import shutil
from unittest.mock import MagicMock
import requests
import requests_mock

import cdiquerygenestoterm
from cdiquerygenestoterm import cdiquerygenestotermcmd


class TestCdgprofilergenestoterm(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_read_inputfile(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('hellothere')
            res = cdiquerygenestotermcmd.read_inputfile(tfile)
            self.assertEqual('hellothere', res)
        finally:
            shutil.rmtree(temp_dir)

    def test_parse_args(self):
        myargs = ['inputarg']
        res = cdiquerygenestotermcmd._parse_arguments('desc',
                                                      myargs)
        self.assertEqual('inputarg', res.input)
        self.assertEqual('http://public.ndexbio.org', res.url)
        self.assertEqual(1, res.polling_interval)
        self.assertEqual(180, res.retrycount)
        self.assertEqual(30, res.timeout)

    def test_run_iquery_no_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = [tfile]
            theargs = cdiquerygenestotermcmd._parse_arguments('desc',
                                                              myargs)
            try:
                cdiquerygenestotermcmd.run_iquery(tfile,
                                                  theargs)
                self.fail('Expected FileNotFoundError')
            except FileNotFoundError:
                pass
        finally:
            shutil.rmtree(temp_dir)

    def test_run_iquery_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = [tfile]
            theargs = cdiquerygenestotermcmd._parse_arguments('desc',
                                                              myargs)
            res = cdiquerygenestotermcmd.run_iquery(tfile,
                                                    theargs)
            self.assertEqual(None, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_get_best_result_by_singularity_one_entry(self):
        qres = {'sources': [{'results': [{'description': 'somedescription',
                                          'details': {'PValue': 5,
                                                      'similarity': 0.002},
                                          'url': 'someurl',
                                          'nodes': 4,
                                          'hitGenes': ['1', '2']}]}]}
        res = cdiquerygenestotermcmd.get_best_result_by_similarity(qres)
        self.assertEqual('somedescription', res['description'])

    def test_get_best_result_by_singularity_two_entries(self):
        qres = {'sources': [{'results': [{'description': 'somedescription',
                                          'details': {'PValue': 5,
                                                      'similarity': 0.002},
                                          'url': 'someurl',
                                          'nodes': 4,
                                          'hitGenes': ['1', '2']},
                                         {'description': 'somedescription2',
                                          'details': {'PValue': 5,
                                                      'similarity': 0.4},
                                          'url': 'someurl',
                                          'nodes': 4,
                                          'hitGenes': ['1', '2']}
                                         ]}]}
        res = cdiquerygenestotermcmd.get_best_result_by_similarity(qres)
        self.assertEqual('somedescription2', res['description'])
        self.assertEqual(0.4, res['details']['similarity'])

    def test_get_best_result_by_singularity_three_entries(self):
        qres = {'sources': [{'results': [{'description': 'somedescription',
                                          'details': {'PValue': 5,
                                                      'similarity': 0.41},
                                          'url': 'someurl',
                                          'nodes': 4,
                                          'hitGenes': ['1', '2']},
                                         {'description': 'somedescription2',
                                          'details': {'PValue': 5,
                                                      'similarity': 0.4},
                                          'url': 'someurl',
                                          'nodes': 4,
                                          'hitGenes': ['1', '2']},
                                         {'description': 'somedescription3',
                                          'details': {'PValue': 5,
                                                      'similarity': 0.2},
                                          'url': 'someurl',
                                          'nodes': 4,
                                          'hitGenes': ['1', '2']}
                                         ]}]}
        res = cdiquerygenestotermcmd.get_best_result_by_similarity(qres)
        self.assertEqual('somedescription', res['description'])
        self.assertEqual(0.41, res['details']['similarity'])

    def test_get_completed_result_error(self):
        with requests_mock.Mocker() as m:
            m.get('http://foo/integratedsearch/'
                  'v1/mytaskid',
                  status_code=500)
            res = cdiquerygenestotermcmd.\
                get_completed_result('http://foo',
                                     'mytaskid',
                                     'hi')

        self.assertEqual(None, res)

    def test_get_completed_result_successful(self):
        with requests_mock.Mocker() as m:
            m.get('http://foo/integratedsearch/'
                  'v1/mytaskid',
                  json={'hi': 'there'})
            res = cdiquerygenestotermcmd.\
                get_completed_result('http://foo',
                                     'mytaskid',
                                     'hi')

        self.assertEqual({'hi': 'there'}, res)

    def test_wait_for_result_done_immediately(self):
        with requests_mock.Mocker() as m:
            m.get('http://foo/integratedsearch/v1/t/status',
                  json={'progress': 100,
                        'status': 'complete'})
            res = cdiquerygenestotermcmd. \
                wait_for_result('http://foo',
                                't', user_agent='hi')
            self.assertEqual(True, res)

    def test_wait_for_result_error(self):
        with requests_mock.Mocker() as m:
            m.get('http://foo/integratedsearch/v1/t/status',
                  json={'progress': 100,
                        'status': 'error'})
            res = cdiquerygenestotermcmd. \
                wait_for_result('http://foo',
                                't', user_agent='hi')
            self.assertEqual(False, res)

    def test_wait_for_result_eventual_success(self):
        with requests_mock.Mocker() as m:
            m.register_uri('GET', 'http://foo/'
                                  'integratedsearch/v1/t/status',
                           [{'json': {'progress': 50, 'status': ''},
                             'status_code': 200},
                            {'status_code': 500},
                            {'json': {'progress': 100, 'status': 'complete'},
                             'status_code': 200}])

            res = cdiquerygenestotermcmd. \
                wait_for_result('http://foo',
                                't', user_agent='hi',
                                polling_interval=0.001)
            self.assertEqual(True, res)

    def test_wait_for_call_timeout(self):
        with requests_mock.Mocker() as m:
            m.register_uri('GET', 'http://foo/'
                                  'integratedsearch/v1/t/status',

                           [{'exc': requests.exceptions.ConnectTimeout},
                            {'status_code': 500},
                            {'json': {'progress': 100, 'status': 'complete'},
                             'status_code': 200}])

            res = cdiquerygenestotermcmd. \
                wait_for_result('http://foo',
                                't', user_agent='hi',
                                polling_interval=0.001)
            self.assertEqual(True, res)

    def test_wait_for_result_retry_exceeded(self):
        with requests_mock.Mocker() as m:
            m.register_uri('GET', 'http://foo/'
                                  'integratedsearch/v1/t/status',
                           [{'json': {'progress': 50, 'status': ''},
                             'status_code': 200},
                            {'status_code': 500},
                            {'json': {'progress': 100, 'status': 'complete'},
                             'status_code': 200}])

            res = cdiquerygenestotermcmd. \
                wait_for_result('http://foo',
                                't', user_agent='hi',
                                polling_interval=0.001,
                                retrycount=2)
            self.assertEqual(False, res)

    def test_get_result_in_mapped_term_json_errors(self):
        # try None
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(None)
        self.assertEqual(None, res)

        # try empty dict
        result = {}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual(None, res)

        # try None sources
        result = {'sources': None}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual(None, res)

        # try empty sources
        result = {'sources': []}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual(None, res)

        # results not in first sources
        result = {'sources': [{'hi': 'there'}]}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual(None, res)

        # results is None
        result = {'sources': [{'results': None}]}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual(None, res)

        # results is empty
        result = {'sources': [{'results': []}]}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual(None, res)

    def test_get_result_in_mapped_term_json_with_colon_success(self):
        result = {'sources': [{'results': [{'description': 'hi: somedescription',
                                            'details': {'PValue': 5, 'similarity': 0.1},
                                            'url': 'someurl',
                                            'nodes': 3,
                                            'hitGenes': ['1', '2']}]}]}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual('somedescription', res['name'])
        self.assertEqual('hi', res['source'])
        self.assertEqual('someurl', res['description'])
        self.assertEqual(5, res['p_value'])
        self.assertEqual(['1', '2'], res['intersections'])
        self.assertEqual(3, res['term_size'])

    def test_get_result_in_mapped_term_json_no_colon_success(self):
        result = {'sources': [{'results': [{'description': 'somedescription',
                                            'details': {'PValue': 5,
                                                        'similarity': 0.001},
                                            'url': 'someurl',
                                            'nodes': 2,
                                            'hitGenes': ['1', '2']}]}]}
        res = cdiquerygenestotermcmd.get_result_in_mapped_term_json(result)
        self.assertEqual('somedescription', res['name'])
        self.assertEqual('NA', res['source'])
        self.assertEqual('someurl', res['description'])
        self.assertEqual(5, res['p_value'])
        self.assertEqual(['1', '2'], res['intersections'])
        self.assertEqual(2, res['term_size'])

    def test_successful_run(self):
        temp_dir = tempfile.mkdtemp()
        try:
            inputfile = os.path.join(temp_dir, 'input.txt')
            with open(inputfile, 'w') as f:
                f.write('hi,there\n')
            user_agent = 'cdiquerygenestoterm/' + cdiquerygenestoterm.__version__
            with requests_mock.Mocker() as m:
                qres = {'sources': [{'results': [{'description': 'somedescription',
                                                  'details': {'PValue': 5,
                                                              'similarity': 0.002},
                                                  'url': 'someurl',
                                                  'nodes': 4,
                                                  'hitGenes': ['1', '2']}]}]}
                m.get('http://foo/integratedsearch/v1/t',
                      json=qres, complete_qs=True)
                m.get('http://foo/integratedsearch/v1/t/status',
                      json={'progress': 100,
                            'status': 'complete'})

                m.post('http://foo/integratedsearch/v1/',
                       request_headers={'Content-Type': 'application/json',
                                        'User-Agent': user_agent},
                       status_code=202, json={'id': 't'})
                myargs = [inputfile, '--url', 'http://foo']
                p = cdiquerygenestotermcmd._parse_arguments('desc',
                                                            myargs)
                res = cdiquerygenestotermcmd.run_iquery(inputfile, p)
                self.assertEqual('somedescription', res['name'])
                self.assertEqual('NA', res['source'])
                self.assertEqual(5, res['p_value'])
                self.assertEqual('someurl', res['description'])
                self.assertEqual(['1', '2'], res['intersections'])
                self.assertEqual(4, res['term_size'])

        finally:
            shutil.rmtree(temp_dir)

    def test_request_failed(self):
        temp_dir = tempfile.mkdtemp()
        try:
            inputfile = os.path.join(temp_dir, 'input.txt')
            with open(inputfile, 'w') as f:
                f.write('hi,there\n')
            user_agent = 'cdiquerygenestoterm/' + cdiquerygenestoterm.__version__
            with requests_mock.Mocker() as m:
                m.post('http://foo/integratedsearch/v1/',
                       request_headers={'Content-Type': 'application/json',
                                        'User-Agent': user_agent},
                       status_code=404, json={'id': 't'})
                myargs = [inputfile, '--url', 'http://foo']
                p = cdiquerygenestotermcmd._parse_arguments('desc',
                                                            myargs)
                res = cdiquerygenestotermcmd.run_iquery(inputfile, p)
                self.assertEqual(None, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_request_wait_failed(self):
        temp_dir = tempfile.mkdtemp()
        try:
            inputfile = os.path.join(temp_dir, 'input.txt')
            with open(inputfile, 'w') as f:
                f.write('hi,there\n')
            user_agent = 'cdiquerygenestoterm/' + cdiquerygenestoterm.__version__
            with requests_mock.Mocker() as m:
                m.register_uri('GET', 'http://foo/'
                                      'integratedsearch/v1/t/status',
                               [{'json': {'progress': 50, 'status': ''},
                                 'status_code': 200},
                                {'status_code': 500},
                                {'json': {'progress': 50, 'status': ''},
                                 'status_code': 200}])
                m.post('http://foo/integratedsearch/v1/',
                       request_headers={'Content-Type': 'application/json',
                                        'User-Agent': user_agent},
                       status_code=202, json={'id': 't'})
                myargs = [inputfile, '--url', 'http://foo',
                          '--polling_interval', '0.001',
                          '--retrycount', '2']
                p = cdiquerygenestotermcmd._parse_arguments('desc',
                                                            myargs)
                res = cdiquerygenestotermcmd.run_iquery(inputfile, p)
                self.assertEqual(None, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_invalid_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = ['prog', tfile]
            res = cdiquerygenestotermcmd.main(myargs)
            self.assertEqual(2, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = ['prog', tfile]
            res = cdiquerygenestotermcmd.main(myargs)
            self.assertEqual(0, res)
        finally:
            shutil.rmtree(temp_dir)


if __name__ == '__main__':
    sys.exit(unittest.main())


cdiquerygenestoterm/cdiquerygenestotermcmd.py
#!/usr/bin/env python

import os
import sys
import argparse
import json
import requests
import time
import cdiquerygenestoterm

SOURCES_KEY = 'sources'
RESULTS_KEY = 'results'
DETAILS_KEY = 'details'
SIMILARITY_KEY = 'similarity'


def _parse_arguments(desc, args):
    """
    Parses command line arguments
    :param desc:
    :param args:
    :return:
    """
    help_fm = argparse.ArgumentDefaultsHelpFormatter
    parser = argparse.ArgumentParser(description=desc,
                                     formatter_class=help_fm)
    parser.add_argument('input',
                        help='comma delimited list of genes in file')
    parser.add_argument('--url', default='http://public.ndexbio.org',
                        help='Endpoint of REST service')
    parser.add_argument('--polling_interval', default=1,
                        type=float, help='Time in seconds to'
                                         'wait between '
                                         'checks on task '
                                         'completion')
    parser.add_argument('--timeout', default=30,
                        type=int, help='Timeout for http '
                                       'requests in seconds')
    parser.add_argument('--retrycount', default=180, type=int,
                        help='Number times to check for completed'
                             'request. Take this value times'
                             'the --polling_interval to determine'
                             'how long this tool will wait'
                             'for a completed result')
    return parser.parse_args(args)


def read_inputfile(inputfile):
    """

    :param inputfile:
    :return:
    """
    with open(inputfile, 'r') as f:
        return f.read()


def get_completed_result(resturl, taskid, user_agent,
                         timeout=30):
    """

    :param resultasdict:
    :return:
    """
    res = requests.get(resturl + '/integratedsearch/v1/' + taskid +
                       '',
                       headers={'Content-Type': 'application/json',
                                'User_agent': user_agent},
                       timeout=timeout)
    if res.status_code != 200:
        sys.stderr.write('Received http error: ' +
                         str(res.status_code) + '\n')
        return None
    return res.json()


def wait_for_result(resturl, taskid, user_agent, polling_interval=1,
                    timeout=30,
                    retrycount=180):
    """
    Polls **resturl** with **taskid**
    :param resturl:
    :param taskid:
    :param user_agent:
    :param polling_interval:
    :param timeout:
    :param retrycount:
    :return: True if task completed successfully False otherwise
    :rtype: bool
    """
    counter = 0
    while counter < retrycount:
        try:
            res = requests.get(resturl + '/integratedsearch/v1/' +
                               taskid + '/status',
                               headers={'Content-Type': 'application/json',
                                        'User_agent': user_agent},
                               timeout=timeout)

            if res.status_code is 200:
                jsonres = res.json()
                if jsonres['progress'] == 100:
                    if jsonres['status'] != 'complete':
                        sys.stderr.write('Got error: ' + str(jsonres) + '\n')
                        return False
                    return True
            else:
                sys.stderr.write('Received error : ' +
                                 str(res.status_code) +
                                 ' while polling for completion')
        except requests.exceptions.RequestException as e:
            sys.stderr.write('Received exception waiting for task'
                             'completion: ' + str(e))

        counter += 1
        time.sleep(polling_interval)
    return False


def get_best_result_by_similarity(resultasdict):
    """
    Gets best result by cosine similarity. The service
    currently sorts by pvalue, but this is not returning
    the best result so lets use cosine similarity

    :param resultasdict:
    :type resultasdict: dict
    :return: best result as a dict
    :rtype: dict
    """
    best_similarity = -1.0
    best_result = None
    for cursource in resultasdict[SOURCES_KEY]:
        for curresult in cursource[RESULTS_KEY]:
            if best_similarity == -1.0 or \
               curresult[DETAILS_KEY][SIMILARITY_KEY] > best_similarity:
                best_result = curresult
                best_similarity = curresult[DETAILS_KEY][SIMILARITY_KEY]
    return best_result


def get_result_in_mapped_term_json(resultasdict):
    """

    :param resultasdict:
    :return:
    """

    if resultasdict is None:
        sys.stderr.write('Results are None\n')
        return None

    if SOURCES_KEY not in resultasdict:
        sys.stderr.write('No sources found in results\n')
        return None

    if resultasdict[SOURCES_KEY] is None:
        sys.stderr.write('Source is None\n')
        return None

    if len(resultasdict[SOURCES_KEY]) <= 0:
        sys.stderr.write('Source is empty\n')
        return None

    if RESULTS_KEY not in resultasdict[SOURCES_KEY][0]:
        sys.stderr.write('Results not in source\n')
        return None

    if resultasdict[SOURCES_KEY][0][RESULTS_KEY] is None:
        sys.stderr.write('First result is None')
        return None

    if len(resultasdict[SOURCES_KEY][0][RESULTS_KEY]) <= 0:
        sys.stderr.write('No result found\n')
        return None

    bestresult = get_best_result_by_similarity(resultasdict)

    colon_loc = bestresult['description'].find(':')
    if colon_loc == -1:
        source = 'NA'
    else:
        source = bestresult['description'][0:colon_loc]

    theres = {'name': bestresult['description'][colon_loc + 1:].lstrip(),
              'source': source,
              'p_value': bestresult['details']['PValue'],
              'description': bestresult['url'],
              'term_size': bestresult['nodes'],
              'intersections': bestresult['hitGenes']}

    return theres


def run_iquery(inputfile, theargs):
    """
    todo

    :param inputfile:
    :param theargs:
    :param gprofwrapper:
    :return:
    """
    genes = read_inputfile(inputfile)
    genes = genes.strip(',').strip('\n').split(',')
    if genes is None or (len(genes) == 1 and len(genes[0].strip()) == 0):
        sys.stderr.write('No genes found in input')
        return None
    user_agent = 'cdiquerygenestoterm/' + cdiquerygenestoterm.__version__
    resturl = theargs.url

    query = {'geneList': genes,
             'sourceList': ['enrichment']}
    res = requests.post(resturl + '/integratedsearch/v1/',
                       json=query, headers={'Content-Type': 'application/json',
                                            'User-Agent': user_agent},
                        timeout=theargs.timeout)
    if res.status_code != 202:
        sys.stderr.write('Got error status from service: ' + str(res.status_code) + ' : ' + res.text + '\n')
        return None

    taskid = res.json()['id']

    if wait_for_result(resturl, taskid, user_agent,
                       timeout=theargs.timeout,
                       retrycount=theargs.retrycount,
                       polling_interval=theargs.polling_interval) is False:
        return None

    resjson = get_completed_result(resturl, taskid, user_agent,
                                   timeout=theargs.timeout)
    return get_result_in_mapped_term_json(resjson)


def main(args):
    """
    Main entry point for program

    :param args: command line arguments usually :py:const:`sys.argv`
    :return: 0 for success otherwise failure
    :rtype: int
    """
    desc = """
        Running gene enrichment against Integrated Query

        Takes file with comma delimited list of genes as input and
        uses Integrated Query (iQuery) enrichment service to 
        to find best network that best matches query genes. The
        best network is the one with highest similarity score.
        This tool then uses the network name as the term name.
        
        The result is sent to standard out in JSON format 
        as follows:
        
        {
         "name":"<TERM NAME WHICH IS NAME OF NETWORK>",
         "source":"<SOURCE OF NETWORK>",
         "p_value":<PVALUE>,
         "description":"<URL OF NETWORK IN NDEx>",
         "term_size":<NUMBER OF NODES IN NETWORK>,
         "intersections":["<LIST OF FOUND IN NETWORK>"]
        }
        
        NOTE: term_size is set to number of nodes in network
              and NOT number of genes
    """

    theargs = _parse_arguments(desc, args[1:])

    try:
        inputfile = os.path.abspath(theargs.input)
        theres = run_iquery(inputfile, theargs)
        if theres is None:
            sys.stderr.write('No terms found\n')
        else:
            json.dump(theres, sys.stdout)
        sys.stdout.flush()
        return 0
    except Exception as e:
        sys.stderr.write('Caught exception: ' + str(e))
        return 2


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))
